### DISCUSSION TITLE
allow multiple predicates for logica_lib.RunPredicateFromString

### DISCUSSION THREAD

**Discussion by wdduncan** [EvgSkv/logica#392] [Ideas]
Perhaps modify `logica_lib.RunPredicateFromString` to allow multiple predicates. E.g.:
```sql
logica_lib.RunPredicateFromString("""
@Engine("sqlite");

Test1(x) :- ....;
Test1(y) :- ....;   
""", ['Test1', 'Test2'])
```

**Comment by EvgSkv** [comment]
I thought about it. It feels to me it would be redundant to have a loop inside the function. Isn't it just as readable to say:

```
for p in ['Test1', 'Test2']:
  logica_lib.RunPredicateFromString("""
  @Engine("sqlite");

  Test1(x) :- ....;
  Test1(y) :- ....;   
  """, p)
```

Or are you expecting the function to optimize the execution in some way?

Thank you for feedback!
Evgeny.

**Reply by wdduncan** [reply]
Ah ... I didn't think about a loop :)  
I would have to try it and see what the output looks like on local Jupyter instance. E.g., would only the last predicate in the list get printed?

**Comment by EvgSkv** [comment]
Just curious, if you are on Jupyter notebook, why are you not using `%%logica` magic?

Like
```
%%logica P, Q

P("P");
Q("Q");
```
?

But anyway, if you want to display results in a loop, then you can use IPython.display.display function:
![image](https://github.com/user-attachments/assets/7e0f0b8e-27ae-4ee3-829a-2aab19b8888c)

**Reply by wdduncan** [reply]
I was not able figure out how to get the `%%logica magic` magic to work on my local jupyter. My jupyter foo is weak :(  
But, I am also using `logica_lib.RunPredicateFromString` to make function calls from within a python library.

**Reply by EvgSkv** [reply]
Hmmm... Did you try adding a softlink (like you mentioned in the other thread) in the folder from which you run your jupyter and calling the
```
%%logica Q
@Engine("sqlite");
Q("Hi!");
```
magic?
What message does it give you?

Magics are nice!

**Reply by wdduncan** [reply]
I get  the error message:
```python
UsageError: Cell magic `%%logica` not found.
```

**Reply by wdduncan** [reply]
BTW, I can do some hacky things with `colab_logica.py`.
```python
from IPython.core.magic import register_cell_magic
from logica.colab_logica import Logica
@register_cell_magic
def logica(line, cell):
  Logica(line, cell, run_query=True)
```
The line `from logica.colab_logica import Logica` produces:
![image](https://github.com/user-attachments/assets/016f26a3-10eb-4b89-98c8-f5a9a36f16eb)

When I run:
```
%%logica Q
@Engine("sqlite");
Q("Hi!");
```
The output is:
![image](https://github.com/user-attachments/assets/5b7fe9f5-4f00-41ac-b995-18af9b3d590b)

**Reply by EvgSkv** [reply]
Alright, so this is exactly how it's supposed to work 😁 
Messages about not being able to import BQ are fine, it's just fyi that it won't be available.
You can list multiple predicates in the cell and they will all render.
